components/butterfly/src/rumor.rs

```diff
@@ -26,23 +26,162 @@
 use std::collections::HashMap;
 use std::default::Default;
 use std::sync::{Arc, RwLock};
+use std::sync::atomic::{AtomicUsize, Ordering};
 
-use member::UuidSimple;
+use message::swim::Rumor_Type;
+use error::{Result, Error};
 
 /// The description of a `RumorKey`.
-#[derive(Clone, Debug, Hash, PartialEq, PartialOrd, Eq)]
+///
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
 pub struct RumorKey {
-    pub kind: String,
-    pub key: UuidSimple,
+    pub kind: Rumor_Type,
+    pub id: String,
+    pub key: String,
 }
 
 impl RumorKey {
-    pub fn new<A: Into<String>, B: Into<String>>(kind: A, key: B) -> RumorKey {
+    pub fn new<A: Into<String>, B: Into<String>>(kind: Rumor_Type, id: A, key: B) -> RumorKey {
         RumorKey {
-            kind: kind.into(),
+            kind: kind,
+            id: id.into(),
             key: key.into(),
         }
     }
+
+    pub fn key(&self) -> String {
+        if self.key.len() > 0 {
+            format!("{}-{}", self.id, self.key)
+        } else {
+            format!("{}", self.id)
+        }
+    }
+}
+
+/// A representation of a Rumor; implemented by all the concrete types we share as rumors. The
+/// exception is the Membership rumor, since it's not actually a rumor in the same vein.
+pub trait Rumor {
+    fn kind(&self) -> Rumor_Type;
+    fn key(&self) -> &str;
+    fn id(&self) -> &str;
+    fn merge(&mut self, mut other: Self) -> bool;
+    fn write_to_bytes(&self) -> Result<Vec<u8>>;
+}
+
+impl<'a, T: Rumor + Clone> From<&'a T> for RumorKey {
+    fn from(rumor: &'a T) -> RumorKey {
+        RumorKey::new(rumor.kind(), rumor.id(), rumor.key())
+    }
+}
+
+/// Storage for Rumors. It takes a rumor and stores it according to the member that produced it,
+/// and the service group it is related to.
+///
+/// Generic over the type of rumor it stores.
+#[derive(Debug, Clone)]
+pub struct RumorStore<T: Rumor> {
+    pub list: Arc<RwLock<HashMap<String, HashMap<String, T>>>>,
+    pub update_counter: Arc<AtomicUsize>,
+}
+
+impl<T: Rumor + Clone> Default for RumorStore<T> {
+    fn default() -> RumorStore<T> {
+        RumorStore {
+            list: Arc::new(RwLock::new(HashMap::new())),
+            update_counter: Arc::new(AtomicUsize::new(0)),
+        }
+    }
+}
+
+impl<T: Rumor + Clone> RumorStore<T> {
+    /// Create a new RumorStore for the given type. Allows you to initialize the counter to a
+    /// pre-set value. Useful mainly in testing.
+    pub fn new(counter: usize) -> RumorStore<T> {
+        RumorStore { update_counter: Arc::new(AtomicUsize::new(counter)), ..Default::default() }
```

Comment:
Love this syntax, I didn't know about it until reading here `{ thing: T, ..Default::default() }`